### DISCUSSION TITLE
Is it possible to port to CYD with SPI ST7789 and a Ra-02 SX1278?

### DISCUSSION THREAD

**Discussion by eben80** [dl9rdz/rdz_ttgo_sonde#481] [Developers corner]
Would it be possible to get the firmware to work with an SPI TFT based board?
https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display/blob/main/DisplayConfig/CYD2USB/User_Setup.h
If i understand correctly it looks like it only supports i2c ST7789 under hardware configuration.
I know that the biggest issue would be to get the free SPI GPIOs to connect the SX1278 but I suppose the SD Card SPI could be cannibalized on hardware level. Iv'e flashed the firmware to the CYD and it starts up with the AP but of course the display doesn't work  since I cannot set up the SPI pins for it. 

Is there anything I could try?

**Comment by dl9rdz** [comment]
Hi,

The support for ILI9225/9341/9342/ST7789 is all based on SPI connections, so this should be fine.

I just took a look at the CYD. The Display seems to be on HSPI and the SDCARD on VSPI.

Usually the software assumes the SX1278 to be on the VSPI, so as long as you don't use the SD-Card this should be fine.
Note that "DC" is the same as "RS" in the config, and MISO is not needed/not present in config.


In theory the SPI bus can be shared between devices (with different chip select pins).
- Sharing between Display and SX1278 works fine on the M5Stack. However, there is this part:
https://github.com/dl9rdz/rdz_ttgo_sonde/blob/331fa13330800cf9cfa609ab7ca725564e6f300f/RX_FSK/src/Display.cpp#L520
where the use of the SPI bus (vspi or hspi) is hard coded. The SX1278 is always supposed to be on the VSPI and the display usually is on the HSPI, but VSPI is used if a M5Stack is detected (this is not configurable in the config at the moment).
- Sharing between Display and SD-Card has some issues on my test setup that I need to debug (TODO...)
- Sharing between SD-Card and SX1278 is untested so far and probably will not work. Synchronization missing

**Reply by eben80** [reply]
Thank you for the response. OK so the Display doesn't work but let me try and get the lora working first. 
For LORA:
Do I understand correctly that I only need to connect MISO(GPIO19), MOSI(GPIO23), SCK(GPIO18), NSS(GPIO5) and 3V3 and Ground?
DIO0 and RST are not connected and configurable?

For TFT:
Display Type = 5
OLED SDA/TFT SDA = MOSI/SDO = GPIO12
OLED SCL/TFT CLK = SCK = GPIO14
OLED RST/TFT RST = -1
TFT RS = GPIO2
TFT CS = GPIO15

The TFT backlight is GPIO21, how do I configure that?
The Green colour on RGB is GPIO16

I'm getting this in logs:

> [20:57:37]Setup sx1278
> [20:57:37]Configuing SX1278FSK SPI with miso=19, mosi=23, sck=18, ss=5
> [20:57:37]0
> [20:57:37]Setting freq: SUCCESS 
> [20:57:37]Frequency set to 0.00
> [20:57:38]0
> [20:57:38]Setting freq: SUCCESS 
> [20:57:38]Frequency set to 0.00
> [20:57:38]RX LNA Gain is -999
> [20:57:38]SX1278 configuration finished
> [20:57:38]Setup finished
> [20:57:38]
> [20:57:38]Sonde.cpp: Invalid rxtask.currentSonde: -1
> [20:57:38]0
> [20:57:38]Setting SX1278 power on FAILED

And then this in a loop:

> [20:57:50]Setting SX1278 power on FAILED
> [20:57:50]TOO MUCH DATA
> [20:57:50]0

The display is not showing anything.

Any advice please?

**Comment by dl9rdz** [comment]
For the backlight you could use the "Power output control". Set it to 21 to pull it low, set it to 149 (21+128) to pull it high - I did not check what the display backlight needs... probably high?

According to https://randomnerdtutorials.com/esp32-cheap-yellow-display-cyd-pinout-esp32-2432s028r/
MOSI is Pin 13, not Pin 12

SX1278:
This should work this way, yes.
The chip will do a power-on reset automatically, on the other boards I do not do any reset in software, and that is fine. The reset pin needs to be high (and this should be ok if it is just left open I guess, but to be sure a pull up to 3V3 should also be fine). 
DIO pins are not used

**Reply by eben80** [reply]
Aah yes, setting it to 149 worked and correcting the MOSI error brought the screen to life and it is displaying properly. I also dropped the SPI speed to 20MHz as it was suggested in the CYD docs.

I'm still having an issue with the Lora though. 

It is still looping with:

> [20:57:50]Setting SX1278 power on FAILED
> [20:57:50]TOO MUCH DATA
> [20:57:50]0

Even after pulling RST to 3.3V

What triggers this TOO MUCH DATA?

**Reply by dl9rdz** [reply]
Technically it means that more than 512 bytes were read from the SX1278 FIFO in packet mode (which is used for RS41). During this time, the SX1278 IRQ FLAGS2 continuously indicates that the FIFO is not empty, but also that the payload is not ready. This should never happen (for RS41, the SX1278 is programmed to received a max 320 byte frame, so after 320 bytes it should indicate frame complete).

Practically, this usually means that all data read from the bus is 0, as then the FIFO empty bit is 0 (=not empty), and the payload ready bit is also 0 (=not ready).

**Comment by RadioAnorak** [comment]
I am facing the same issue here with the Ra-02 module and receive this serial output after assigning SPI pins in the config.

```
01:55:45.334 -> 0
01:55:45.334 -> Setting SX1278 power on FAILED
01:55:45.379 -> [0;32mSonde.cpp: Sonde::setup() done: Type RS41 Freq 0.000000, AFC BW: 500000, RX BW: 500000[0m
01:55:45.379 -> TOO MUCH DATA
```

Would there possibly be a slight inherent difference between the Ra-02 module and SX1278 that is not allowing the communication to work? Even though their operation is practically identical?

Many thanks.

**Reply by hr-ru** [reply]
There should not be any relevant difference.
"Setting SX1278 power on FAILED" basically shows that the communication with the SX1278 is not working.
The "TOO MUCH DATA" is just a side effect that communication not working: the software tries to read as much data as there is in the chip's FIFO (until a control register of the chip shows that the FIFO is empty). If the communication is not working, then this reading of the control register may ALWAYS indicate "not empty", which after reading more data than the FIFO should contain triggers the "TOO MUCH DATA" message.  

Possible issues
- Wrong Pin assignments/configuration (obvious)
- Duplicate Pin assignments (check if your SX1278 pins are not additionally configured somewhere else, like LED out, SD-card, etc.)
- Maybe a reset issue? The software does not manage the SX1278 reset pin and solely relies on the chip being reset internally on power up (this should work according to the data sheet). The reset pin should be floating/High-Z for that purpose.
- Maybe your custom board is mis-detected as an M5Stack? Let me know if the "About" tab shows your board as that type.

**Reply by RadioAnorak** [reply]
Thank you very much for your reply. That's what I assumed and find very strange, albeit I am using a custom hardware setup on a breadboard for now. The Ra-02 I am using is known to work for other projects.

I am using a custom setup with a Wemos Lolin32 and have managed to set the config settings such that the built in screen works and displays the startup message.

The SPI pins I have chosen for the board I am certain are correct in the config and have checked multiple times to make sure I was not being silly. 

The pins have not been duplicated, for LED out and SD card pins I have assigned them as -1 as I solely care about receiving packets for now.

Reset may be the issue? The pin is floating with no pin connected so I would hope that is correct.

It has autodetected my board as a TTGO LORA v1.0 (0.9" OLED@4,15) on the about section.

**Reply by hr-ru** [reply]
The only place I have used the Ra-02 is the M5Stack Lora v1.0 module (https://docs.m5stack.com/en/module/lora), and this is working fine. So basically a Ra-02 should work.

Regarding reset, I did take another look: On the M5Stack Gray the Ra-02 Reset is connected to the ESP32 GPIO26, which is not configured by the software, so this should be in the default "disabled" state.  This is similar to the other boards, e.g. the T-Beam has a hardware connection of the SX1278's RESET to GPIO23 (again, not configured in software, so also in default state).

Which pins are you using for the Ra-02?

**Reply by RadioAnorak** [reply]
That sounds promising! Sounds like a me issue then with the hardware setup.

For this ESP32 board, it doesn't have many free GPIO pins - at the moment I am using the following:

- NSS: 15
- MOSI: 13
- MISO: 12
- SCK: 14

RST is still not connected. I just tried with some 'floating' pins to get a high Z but still with no joy. I have tried my 25 and 26 pins too and the same result.

Here is the board I am using: [https://randomnerdtutorials.com/esp32-built-in-oled-ssd1306/](https://randomnerdtutorials.com/esp32-built-in-oled-ssd1306/) 

Pins 4 and 5 are used for I2C for the screen which when set in the config makes it work.

I do have access to other ESPs I will try but they do not have screens built in.

**Comment by d2f2** [comment]
Just got cyd 2.8 and have some problem with tft.
All texts are drawn normalne but sourronded by White background.

Display type (0=OLED/SSD1306, 1=ILI9225, 2=OLED/SH1106, 3=ILI9341, 4=ILI9342, 5=ST7789, 6=ST7796)	
5

OLED SDA/TFT SDA	
13

OLED SCL/TFT CLK	
14

OLED RST/TFT RST (needs reboot)	
-1

TFT RS	
2

TFT CS	
15

TFT SPI speed	
40000000
(tested at 10000000, 20M, 40M no difference) 


Button input port	
-1
 Touch
Button 2 input port	
-1
 Touch
Use AXP192 PWR as Button 2	
0
Touch button threshold
(0 for calib mode)	
70
Power control port	
149


Trying ili9341 driver caused colors reverse but problem occurs.

**Comment by d2f2** [comment]
![20251106_213038](https://github.com/user-attachments/assets/bbca3357-cc25-4319-8eeb-8bc47412ff0a)
![20251106_212712](https://github.com/user-attachments/assets/6b7f2e77-8f29-49c0-9649-a672f536223d)
![20251106_211332](https://github.com/user-attachments/assets/5c84f626-4dee-4ba8-9487-66927f6ea9a1)
![20251106_213147](https://github.com/user-attachments/assets/c747d591-ac14-48ab-a741-7fda30f37cae)

**Comment by dl9rdz** [comment]
According to https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display/blob/main/OriginalDocumentation/4-Driver_IC_Data_Sheet/JC2432A028N.pdf the display chip should be an ILI9341V, which seems to be basically the same as a ILI9341.

Maybe you need to reduce SPI clock speed, 40 MHz is out of spec for the ILI9341. According to data sheet write clock cycle length is 100ns so max 10 MHz. Practically I am using an ILI9341 with up to 20 MHz without problems, but 40 MHz does not work reliably. Maybe this helps with your problem?

**Comment by d2f2** [comment]
I have tested at 10M, 20M, 40M.
There is no fix at 10M. Just slower drawing.

9341 driver works the same but negative colors.

I think i have had same problem with other setup wroom32+ili, but cant renember solutions. Have Compared settings, but only pins are diffrent due to hardware setup. Also at working setup with other ili tft have tft RST connected and assingned. Here cyd have
OLED RST/TFT RST (needs reboot)
-1

**Comment by hr-ru** [comment]
So I now got one board for testing, this one: https://www.amazon.es/dp/B0DXFBKKQB

With that board, everything works nicely with 3/ILI9341 as type (with 5/ST7789 the display remains blank white). However, there also seems to be a different board available with ST7789 (https://www.amazon.es/dp/B0FFZ52YV5). 

That said, your board seems to be yet another board (you have this R21 next to the display which does not exist on my board and also is not visible on the photos of the other board I linked above).

**Comment by dl9rdz** [comment]
Some more experiments, I now got the another board with ST7789 as well, and displays works just fine. 
The boards I have seem to be some newer flavour with USB-C port. The nice thing about both these boards is that they have two 4-pin connectors, one with 4 IO pins for SPI and one with GND, 3V3 and an input, so this would be ideal to connect a SX1278 (Ra-02) and a GPS RX (and there is one free pin in the second connector, one could wire a GPS TX wire if one wants to - not strictly needed though)

<img width="973" height="550" alt="Screenshot 2025-11-22 at 23 53 25" src="https://github.com/user-attachments/assets/0112aba1-9896-4b88-b345-efc5f937e942" />

**Reply by dl9rdz** [reply]
<img width="1002" height="535" alt="Screenshot 2025-11-22 at 23 58 32" src="https://github.com/user-attachments/assets/97f766f1-0dad-45d9-861d-1c5ea011df97" />

**Comment by NEXA-USA** [comment]
Hi All. I'm going to layout a dedicated version of this with the RA-02 included.  I'm new to Git and not sure if the final configuration of this ended up being.  Which LCD is final, or if there are multiple parallel configurations? Is this still an ongoing effort? Is there code I could test with? Any advice/guidance would be very helpful.
